tensorflow/core/distributed_runtime/rpc/grpc_remote_worker.cc

```diff
@@ -260,10 +288,206 @@ class GrpcRemoteWorker : public WorkerInterface {
+    // std::cerr << response->DebugString() << std::endl;
+    RTresponse.ParseFromArray(&sta[4], sta.size() - sizeof(int));
+    // std::cerr << RTresponse.DebugString() << std::endl;
+
+    // Initialize the destination tensor
+    // TODO is it possible that this tensor already exists/reuse of memory
+    // location?
+    response->InitPartial(RTresponse);
+    const size_t nBytes = response->tensor().TotalBytes();
+
+    // Receive the Tensor content
+    char* data = const_cast<char*>(response->tensor().tensor_data().data());
+    for (size_t i = 0; i <= nBytes; i += maxSize) {
+      const int toRecv = std::min(maxSize, nBytes - i);
+      MPICheck(MPI_Recv(&data[i], toRecv, MPI_BYTE, src, followUpTag,
+                        MPI_COMM_WORLD, &status));
+    }
+#endif
+  }  // MPIRecvTensor
+
+  // Different MPI implementations use different values for allowed tag ranges
+  // so retrieve the max value to generate valid tags
+  int maxMessageTag;
+
+  // Max size of the data chuncks, 512MB
+  const size_t maxSize = 1024 * 1024 * 512;
+
+  // If the MPI_PATH_DISABLED variable is set then disable this code path
+  bool disabledAtRuntime = false;
+
+ public:
+  explicit RDMARemoteWorker(SharedGrpcChannelPtr channel,
+                            ::grpc::CompletionQueue* completion_queue,
+                            WorkerCacheLogger* logger)
+      : GrpcRemoteWorker(channel, completion_queue, logger) {
+    // Determine the maximum allowed message tag, used to determine hash
+    void* v;
+    int flag;
+    MPICheck(MPI_Comm_get_attr(MPI_COMM_WORLD, MPI_TAG_UB, &v, &flag));
+    maxMessageTag = *(int*)v;
+
+    // Test if this path is disabled at launch
+    const char* env = getenv("MPI_PATH_DISABLED");
+    if (env && env[0] == '1') disabledAtRuntime = true;
+  }
+
+  void SendTensorSync(const WorkerEnv* env, const Rendezvous::ParsedKey& key,
+                      const Rendezvous::Args& args, const Tensor& val,
+                      const bool is_dead, const int64 step_id_,
+                      Status& s) override {
+    const int dst = getMPIPartnerID(false, key.FullKey().ToString(), env);
+    const int hash =
+        tensorHash(strings::StrCat(key.FullKey().ToString(), step_id_));
+    // fprintf(stderr, "JBDBG Going to send data to: %s  : %d  || %d
+    // devContext: %d\n", key.FullKey().ToString().c_str(), dst, hash,
+    // args.device_context);
+    MPISendTensor(dst, hash, is_dead, val);
+    s = Status::OK();
+  }
+
+  void RecvTensorAsync(WorkerEnv* env, CallOptions* call_opts,
+                       const RecvTensorRequest* request,
+                       TensorResponse* response, StatusCallback done) override {
+    if (disabledAtRuntime) {
+      GrpcRemoteWorker::RecvTensorAsync(env, call_opts, request, response,
+                                        done);
+    } else {
+      // TODO: Figure out how to get the size of the requested Tensor
+      // Is this known by looking up the key/request?
```

Comment:
You can do this without a round trip if you can pre-post a buffer that is larger than any you will receive, which is basically the eager protocol `MPI_Send` uses internally.  Or you can do the rendezvous approach mentioned already, which is of course also used by `MPI_Send` internally, although there are many different variants of rendezvous depending on what the hardware can do.  Depending on the usage, you can use both, as MPI does internally.

In addition to the obvious rendezvous protocol that uses two send-recv pairs, one can do a variant that uses `MPI_Get` at the target to pull data from the origin, which is how e.g. Blue Gene does rendezvous.  This assumes that `MPI_Get` is asynchronous, which is not always true.  I can work up examples for a bunch of different ways to implement this if people are interested.  Pseudo-code is easy, but I have on occasion written educational code that actually works 😃  ([example](https://github.com/jeffhammond/HPCInfo/blob/master/mpi/active-messages/armci.c))

In the end, what is needed here is an active-message interface, which MPI doesn't expose but often uses internally.  The MPICH team [prototyped](http://www.mcs.anl.gov/~thakur/papers/mpi-active-messages.pdf) an implementation of active-messages that extends the RMA interface, but it is obviously not portable to other MPI implementations.